From dbf208e7e1ad5c03e39c8bdb0c62791102f4146f Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Mon, 30 Sep 2024 13:28:13 +0200
Subject: [PATCH 3/7] media: dcmipp: statcap: correct bins stat capture

In FULL profile, all stats are only available upon capture
of the 4th set of BINS.
Moreover, fix the storage pointer of the 4th set of bins.

Change-Id: Id5b03d72d3950cd5c404da24167346bd7da8e3a5
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
---
 .../media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c
index f261e65c461c..95459569b337 100644
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-statcap.c
@@ -792,9 +792,15 @@ static irqreturn_t dcmipp_statcap_irq_thread(int irq, void *arg)
 				  DCMIPP_P1STXCR_ENABLE);
 
 		if (vcap->prev_capture_state == PHY_BIN_3_SHA_AV_RGB) {
+			/* The data capture refer to the previous location */
+			avr_bins = !vcap->stat_location == DCMIPP_P1STXCR_SRC_LOC_PRE ?
+					&vcap->local_buf.pre : &vcap->local_buf.post;
 			/* Accumulators contains the 4th set of BINS */
 			for (i = 0; i < 3; i++)
 				avr_bins->bins[i + 9] = reg_read(vcap, DCMIPP_P1STXSR(i));
+			/* By the time we get the 4th POST BINS, stat_location is again in PRE */
+			if (vcap->stat_location == DCMIPP_P1STXCR_SRC_LOC_PRE)
+				vcap->stat_ready = true;
 		}
 		break;
 
@@ -873,8 +879,6 @@ static irqreturn_t dcmipp_statcap_irq_thread(int irq, void *arg)
 		if (vcap->capture_state < PHY_BIN_3_SHA_AV_RGB) {
 			vcap->capture_state++;
 		} else {
-			if (vcap->stat_location == DCMIPP_P1STXCR_SRC_LOC_POST)
-				vcap->stat_ready = true;
 			vcap->stat_location = !vcap->stat_location;
 			vcap->capture_state = PHY_AV_RGB_SHA_BIN_0;
 		}
-- 
2.25.1

